Introduce distinct ObjectMappers for RestEasy Reactive / RestEasy Reactive Client

chonton opened the pull request:
Closes #23979

Provide three ObjectMappers: unqualified, client, and server that can be independently produced, customized, and injected.  This allows different behaviors for the different uses.

chonton committed:
23979: Distinct ObjectMappers for RestEasy Reactive / RestEasy Reactive Client

geoand commented:
Thanks for this.

To be honest, I would like to limit the scope of the change.

What I had in mind was to have these separate `ObjectMapper`s only for RESTEasy Reactive and also (and probably more importantly), only leverage the client and server `ObjectMapper` objects **if** they have been created. Otherwise, RESTEasy Reactive should use the default `ObjectMapper`.
The reason is that `ObjectMapper` is very very heavy (both in terms of heap space used and of time taken to create it), so the common use case where users don't need per-case instances should not suffer any performance penalty just to support the less common case.

gsmet commented:
Yes, we need to be extremely careful about that.

kdubb commented:
What do you guys think about using `JsonMapper` instead of `ObjectMapper` during this change? Or maybe in another PR? As of `2.10`, `JsonMapper` appears to be the preferred system.

gsmet commented:
Another thing we need to be careful about is backwards compatibility. You might have people relying on the existing behavior so we need to be careful.

chonton commented:
I've create #24019 to track using `JsonMapper`

chonton commented:
Limited the scope of the change per @geoand 's comments

geoand commented:
Thanks, I'll take a closer look soon.

chonton commented:
@geoand, Rebased and addressed  concerns.  I had to move annotations to `quarkus-resteasy-reactive-jackson-common` module rather than `quarkus-resteasy-reactive-jackson` module.

Are the qualifier annotations in the package that you expect?

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `client` ObjectMapper
```
I think we should also mention what the purpose of this `client` ObjectMapper is - i.e. that it is only used for the reactive rest client

chonton replied:
done

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ServerObjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `server` ObjectMapper
```
Same here

geoand commented on docs/src/main/asciidoc/rest-json.adoc:
```diff
@@ -245,6 +245,149 @@ public class CustomObjectMapper {
 ----
 
+There is one unqualified `ObjectMapper` instances provided by the quarkus-jackson extension.  You may produce two additional
+`ObjectMapper` instances qualified by the `ClientObjectMapper` and `ServerObjectMapper` qualifier annotations.  If provided, the
```
`ClientObjectMapper` -> `@ClientObjectMapper`, `ServerObjectMapper` -> `@ServerObjectMapper`

geoand reviewed (commented):
I added a final round of comments.

We'll also need the commits squashed.

Thanks!

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ClientObjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+package io.quarkus.resteasy.reactive.jackson.common.runtime;
```
Actually, it's probably best if this is moved into the client module.
Also, we don't want to include `runtime` in the package.

chonton replied:
I have moved annotations to `io.quarkus.resteasy.reactive.jackson.common` package.  Annotations remain in `quarkus-resteasy-reactive-jackson-common` module.

Moving annotations to `quarkus-resteasy-reactive-jackson` module causes build to fail with following exception:

Deployment artifact io.quarkus:quarkus-rest-client-reactive-jackson-deployment::jar:999-SNAPSHOT is missing the following dependencies from its configuration: io.quarkus:quarkus-resteasy-reactive-jackson-deployment::jar, io.quarkus:quarkus-resteasy-reactive-deployment::jar, io.quarkus:quarkus-vertx-http-deployment::jar -> [Help 1]

Fixing that causes BasicTest in `integration-tests/rest-client-reactive` module to fail.  This seems way too complicated to fix.

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/runtime/ServerObjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+package io.quarkus.resteasy.reactive.jackson.common.runtime;
```
Same comment as the client

chonton commented on docs/src/main/asciidoc/rest-json.adoc:
```diff
@@ -245,6 +245,156 @@ public class CustomObjectMapper {
 }
 ----
 
+There is one unqualified `ObjectMapper` instance provided by the quarkus-jackson extension.  Your code may produce two additional
+`ObjectMapper` instances. The client `ObjectMapper` instance is qualified by the `@ClientObjectMapper` annotation. The server
+`ObjectMapper` instance is qualified by the `@ServerObjectMapper` annotation.
```
I am happy to change to `@RestClientObjectMapper` and `@RestServerObjectMapper` if that is consensus.

geoand replied:
I like the idea

chonton replied:
Alternatively, a single annotation `@ObjectMapperTarget` with `ObjectMapperAudience` value.  `ObjectMapperAudience` is an extensible enum with current values of `REACTIVE_REST_CLIENT` and `REACTIVE_REST_SERVER`.

chonton replied:
Reverted to `@RestClientObjectMapper` and `@RestServerObjectMapper`
Rebased and squashed.
Any other issues to address?

chonton commented on extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/ClientObjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+@Qualifier
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.TYPE })
+public @interface ClientObjectMapper {
```
moved annotations to io.quarkus.resteasy.reactive.jackson.common

geoand commented:
LGTM, thanks a lot!

@gsmet any more comments on your side?

chonton commented:
@geoand @gsmet Ready to merge?

geoand commented:
Let's wait for Guillaume's comments

gsmet commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/RestClientObjectMapper.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `client` ObjectMapper. If an ObjectMapper
+ * is produced with this qualifier, it is used by the quarkus-jaxrs-client-reactive extension for serialization and
+ * deserialization. Otherwise, quarkus-resteasy-reactive uses the unqualified `ObjectMapper` instance.
```
```suggestion
 * deserialization. Otherwise, quarkus-rest-client-reactive uses the unqualified `ObjectMapper` instance.
```

gsmet commented on docs/src/main/asciidoc/rest-json.adoc:
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
+`ObjectMapper` instances. The client `ObjectMapper` instance is qualified by the `@RestClientObjectMapper` annotation. The server
+`ObjectMapper` instance is qualified by the `@RestServerObjectMapper` annotation.
+
+If the client `ObjectMapper` instance is available, it is used by the quarkus-jaxrs-client-reactive extension for serialization
```
```suggestion
If the REST client `ObjectMapper` instance is available, it is used by the quarkus-rest-client-reactive extension for serialization
```

gsmet commented on docs/src/main/asciidoc/rest-json.adoc:
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
+instance.
+
+If the server `ObjectMapper` is available, it is used by the quarkus-resteasy-reactive extension for serialization and
+deserialization.  If the server instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
```
```suggestion
deserialization.  If the REST server instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
```

gsmet commented on docs/src/main/asciidoc/rest-json.adoc:
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
+You can produce, inject, and customize each of these instances independently.  The following snippet demonstrates how to inject
+each instance using the `@RestClientObjectMapper` and `@RestServerObjectMapper` annotations.
+
+[source,java]
+----
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.quarkus.resteasy.reactive.jackson.common.RestClientObjectMapper;
+import io.quarkus.resteasy.reactive.jackson.common.RestServerObjectMapper;
+
+import javax.inject.Inject;
+
+public class UseFallback {
+
+    private final ObjectMapper serializer;
+
+    @Inject
+    public UseFallback(ObjectMapper mapper,
+            @RestClientObjectMapper Instance<ObjectMapper> clientInstance) {
+        serializer = clientInstance.isUnsatisfied() ? mapper : clientInstance.get();
+    }
+
+}
+----
```
I'm not sure we really want to document this. These mappers are supposed to be used for a very specific purposes by the RESTEasy Reactive extension so I'm not sure I would document anything except the customization of them.

In any case, if we want to keep that (I recommend we don't), I would put it at the end, not at the start.

chonton replied:
Makes sense to remove the fallback in this example.  When would someone wish to inject the alternate ObjectMappers into non-quarkus code?  Possibly/Probably an edge case.

gsmet commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson-common/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/common/RestServerObjectMapper.java:
```diff
@@ -0,0 +1,24 @@
+import javax.inject.Qualifier;
+
+/**
+ * Qualifier Annotation that specializes production, customization, and injection of `server` ObjectMapper. If an ObjectMapper
```
```suggestion
 * Qualifier Annotation that specializes production, customization, and injection of the REST server ObjectMapper. If an ObjectMapper
```

gsmet commented on docs/src/main/asciidoc/rest-json.adoc:
```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
+If you want to customize the unqualified, client, and server instances differently, you can use the `@RestClientObjectMapper` and
+`@RestServerObjectMapper` annotations on your `ObjectMapperCustomizer`.  `ObjectMapperCustomizer` instances without any
+qualifier will customize all three `ObjectMapper`s. `ObjectMapperCustomizer` annotated with `@RestClientObjectMapper` will customize
+only the client instance.  `ObjectMapperCustomizer` annotated with `@RestServerObjectMapper` will customize only the server instance.
```
Hmmmm. With the way it's documented, I would have expected the client and server instances to be created automatically if there are customizers and AFAICS from the code, it's not the case?

I.e. I would have expected the third part of the doc where you have to produce the `ObjectMapper` yourselves to not be necessary as it's quite brittle and you could definitely get it wrong doing things by yourself.
So basically, I would have produced synthetic beans for these object mappers if the annotations were found on customizers as IIRC, it's not possible to return null if we want to produce a `@Singleton`.

@geoand does it make sense or am I missing something?

chonton replied:
Prior comment from some reviewer asked to remove creation of client and server instances.  My quarkus mojo may not be up to the task of creating the beans as part of build.

gsmet replied:
So we don't want to create the ``ObjectMapper``s if they are not needed. My question was: could we create them only if they are required. I think we could do that with synthetic beans but I want to check with @geoand that this makes sense.

gsmet replied:
Hum. And I'm not really sure this works because IIRC the Client writer is only initialized if the server part is not around: https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-client-reactive-jackson/deployment/src/main/java/io/quarkus/rest/client/reactive/jackson/deployment/RestClientReactiveJacksonProcessor.java#L42 .

Also the `JacksonBasicMessageBodyReader` that you haven't changed and should probably be changed is shared between the two.

So my guess is that this needs more thoughts. But I might be missing something as I'm not that familiar with RESTEasy Reactive.

geoand replied:
@gsmet so basically you are asking if we can create the "RestServerObjectMapper" if there are customizers for it?

gsmet replied:
@geoand so that was my initial question yes. As otherwise it's very cumbersome (especially, since you have to apply the customizers yourself).

But... my other question is: I'm pretty sure the current proposal won't work because in some cases we share the writer between client and server and if I'm not mistaken the reader is always shared.
I'm wondering if we could use `@ConstrainedTo(RuntimeType.SERVER)` / `@ConstrainedTo(RuntimeType.CLIENT)` but I'm not sure it's implemented in RESTEasy Reactive.
If we can, we would make the reader/writer for client/server totally distinct and this could work.

geoand replied:
> As otherwise it's very cumbersome (especially, since you have to apply the customizers yourself).

I think it's acceptable price to pay, at least initially until we see if people actually use this feature.

> I'm pretty sure the current proposal won't work because in some cases we share the writer between client and server and if I'm not mistaken the reader is always shared.

What reader and writter are you talking about?

geoand replied:
Good point, I had totally forgot about that! Indeed we would need to always have both and limit them to client and server

gsmet reviewed (changes requested):
I added some minor suggestions but I have also one major concern we need to sort out. See inline.

chonton committed:
Apply suggestions from code review

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

chonton committed:
Apply suggestions from code review

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

geoand commented:
I recently came across another use case where this would be very useful, I think we should reevaluate it, WDYT @gsmet?
